designs/0005-static-matrices.md

```diff
@@ -0,0 +1,240 @@
+[here](https://github.com/stan-dev/math/blob/develop/stan/math/rev/core/vari.hpp#L746)
+
+## Use and Testing in Functions
+
+Because varmat types are meant for performance, there are no default constructors
+for building matvar types from varmat or varmat from matvar (or any of the
+the varmat types above from any other). This is done on purpose to avoid implicit
+conversions and any accidental slowdowns this may lead to. Everything written for
+varmat types is done explicitly, which makes it somewhat less automatic than
+matvar. The matvar types are still available for general purpose autodiff.
+
+Testing varmat autodiff types for different functions is done by including a
+`stan::test::test_ad_varmat(...)` mixed mode autodiff test along with every
+`stan::test::test_ad(...)` test. The tests for `log_determinant`, available
+[here](https://github.com/stan-dev/math/blob/develop/test/unit/math/mix/fun/log_determinant_test.cpp)
+are typical of what is done.
+
+`test_ad_varmat` checks that the values and Jacobian of a function autodiff'd with
+varmat and matvar types behaves the same (and also that the functions throw
+errors in the same places). `test_ad_varmat` is not included by default in `test_ad`
+because not all functions support varmat autodiff.
+
+`test_ad_varmat` also checks return type conventions. For performance, it is assume
+a function takes in a mix of varmat and matvar types should also return a varmat.
+
+A complete list of varmat functions is planned to be developed for the compiler
+implementation. As an incomplete substitute, a list of functions with reverse mode
+implementations that are being converted to work with varmat is here:
+https://github.com/stan-dev/math/issues/2101 . A pull request
+([here](https://github.com/stan-dev/math/pull/2214)) is up that adds support for
+varmat to all the non-multivariate distribution functions.
+
+## Assigning a varmat
+
+Something in the original design doc that is not necessary anymore is the limitation
+that subsets of varmat variables cannot be assigned efficiently.
+
+The efficiency loss was so great in the original design (requiring a copy of the entire
+varmat) that this idea was discarded completely.
+
+There is a strategy available now to assign to subsets of varmat variables that is
+more efficient than this.
+
+The idea is that when part of a varmat is assigned to, the values in that part of the
+varmat are saved into the memory arena before being overwritten, and a call is pushed
+onto the chaining stack that will restore the values back from the arena and set the
+respective adjoints of the varmat entries to zero.
+
+The assumptions that makes this work is after a subset of a varmat is overwritten,
+that part of the varmat cannot be used in any further calculations, and so the adjoints
+of that part of the matrix will be zero in the reverse pass at the time of assignment.
+
+This means that assignment into a varmat would invalidate any views into that
+varmat's memory.
```

Comment:
I would not say invalidates. Just the change is visible in views. Also I don't think you explained anywhere what views are.

Reply:
I still like invalidates, but it's not a segfault or anything to read the view. Perhaps a segfault would be better though since it would prevent inadvertent bugs, so I'm not sure. I changed it.